File: uploader/process/clone.go
package process

import (
	"uploader/model"

	git "github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/storage/memory"
)

func Clone(upload model.Upload) (*git.Repository, error) {
	return git.Clone(memory.NewStorage(), nil, &git.CloneOptions{
		URL: *upload.Repo,
	})
}

File: lib/src/uploader/clone.rs
use crate::db::{upload, Database};
use crate::entities::Upload;
use crate::uploader::{error::Result, Git};

pub async fn clone(db: &Database, upload: &mut Upload, git: &Git) -> Result<Upload> {
    tracing::info!("Cloning repository...");

    git.clone_repo(&format!("/tmp/{}", upload.id), &upload.repo)?;
    upload::set_sha(db, upload.id, git.sha(&format!("/tmp/{}", upload.id))?).await?;

    upload::set_status(db, upload.id, "cloned").await?;
    Ok(upload::append_log(db, upload.id, "INFO: Repository cloned").await?)
}
